bd: Add randomness to block device test and more debug friendly output

geky opened the pull request:
Added to block device test during debug of SPIFBlockDevice and I2CEEBlockDevice. Currently this tests is duplicated for all block devices.

- Uses rand to select pseudorandom blocks, slightly better coverage and no longer will intensively wear block zero.
- Added nice printing of sizes
- Added preview of read/write data for debugging
- Added an error mask of where errors occured in block for debugging

simonqhughes commented:
Please could you take a look at the heap_block_device/main.cpp error at:

https://jenkins-internal.mbed.com/job/ARMmbed/job/mbed-os/job/PR-3747/1/consoleFull

```
[K64F ARM] Build failures:
[K64F ARM]   * K64F::ARM::FEATURES-TESTS-FILESYSTEM-HEAP_BLOCK_DEVICE
[K64F ARM]         Building project heap_block_device (K64F, ARM)
[K64F ARM]         Scan: ARM
[K64F ARM]         Scan: FEATURE_LWIP
[K64F ARM]         Scan: FEATURE_STORAGE
[K64F ARM]         Scan: heap_block_device
[K64F ARM]         Compile [100.0%]: main.cpp
[K64F ARM]         [Error] main.cpp@50,0:  #308: more than one instance of overloaded function "log" matches the argument list:
[K64F ARM]         [Error] main.cpp@50,0:  #308: more than one instance of overloaded function "log" matches the argument list:
[K64F ARM]         
[K64F ARM] [mbed] ERROR: "python" returned error code 1.
[K64F ARM] [mbed] ERROR: Command "python -u /builds/ws/ARMmbed/mbed-os/PR-3747/mbed-os/tools/test.py -t ARM -m K64F -c --source /builds/ws/ARMmbed/mbed-os/PR-3747/mbed-os --build /builds/ws/ARMmbed/mbed-os/PR-3747/mbed-os/BUILD/tests/K64F/ARM --test-spec /builds/ws/ARMmbed/mbed-os/PR-3747/mbed-os/BUILD/tests/K64F/ARM/test_spec.json" in "/builds/ws/ARMmbed/mbed-os/PR-3747/mbed-os"
[K64F ARM] ---

```


@geky

geky commented:
Ah, multiple float/double overloads for the math problems caused inference problems when passed ints.

Should be updated 👍

bridadan commented:
retest uvisor

sg- commented:
/morph test-nightly

0xc0170 commented:
please look at failures. Some seems to be legitimate. I'll rerun it, because some seem to be not relevant to this patch. 

This patch might affect integration of this branch to master #3762

/morph test-nightly

bridadan commented:
/morph test-nightly

0xc0170 commented:
One example for nucleo and for only IAR, please have a look

0xc0170 commented:
Any update for this?

geky commented:
Will update, there's a bit of work to match last minute changes to the block device api

0xc0170 commented:
> Will update, there's a bit of work to match last minute changes to the block device api

Any update?

geky commented:
Updated and rebased against master, although this _will_ have a merge conflict with https://github.com/ARMmbed/mbed-os/pull/3866 because of stupid C++ rules regarding implicit casts of const member function pointers... : (

geky commented:
@bridadan, this also removes the ifdefs preventing tests from running on small devices (no longer an issue).

bridadan commented:
> @bridadan, this also removes the ifdefs preventing tests from running on small devices (no longer an issue).

This makes me a very happy :panda_face:

simonqhughes commented:
Looks good to me. Lets merge.

geky commented:
/morph test-nightly

0xc0170 commented:
Please look at the failures

geky commented:
/morph test

0xc0170 commented:
```
18:03:22         [Error] main.cpp@39,1: cannot convert 'bd_size_t (BlockDevice::*)() const {aka long long unsigned int (BlockDevice::*)() const}' to 'bd_size_t (BlockDevice::*)() {aka long long unsigned int (BlockDevice::*)()}' in initialization
```

All build fails due to error in the test, see above

geky commented:
I _did_ call it, though I forgot this would happen.

Should be updated

geky commented:
/morph test-nightly

bridadan commented:
Results look good!

0xc0170 commented:
@simonqhughes Please review

0xc0170 commented:
> @simonqhughes Please review

Any update?
